### PR TITLE
Merge features/readonly-ref to master

### PR DESCRIPTION
**Pull request by VSadov**
Merge from features/readonly-ref into master.

Brings a number of interconnected features geared towards system/hi-perf programming.

User facing features:
- `ref readonly`  parameters
- `ref readonly` return
- `ref` extension methods
- `ref readonly` extension methods
- `ref` ternary expressions
- `readonly` stucts
- `ref` structs     (such as `Span<T>` and `ReadonlySpan<T>`  and structs that contain them)
- escape analysis for byref variables and `ref struct` values
- safe `stackalloc`
- various optimizations for scenarios that involve readonly fields, stackalloc, fixed, unsafe . . .
- peverify-compat mode

Internal improvements:
- attribute embedding infrastructure
- separation of variable/lvalue and writeable/readonly analysis

==== See also
- Championed feature at https://github.com/dotnet/csharplang/issues/666
- Proposal for `[ref] readonly` at https://github.com/dotnet/csharplang/blob/master/proposals/readonly-ref.md    (ignore `in`, that is gone)  
- Detailed spec for Ref-Like types at https://github.com/dotnet/csharplang/blob/master/proposals/span-safety.md
- Smaller design details of the features above. - https://github.com/VSadov/csharplang/blob/RefSpanDesignChoices/proposals/RefReadonlySpanDesignChoices.md
- Test plan for ref `[ref] readonly` at https://github.com/dotnet/roslyn/issues/19216
- Test Plan for Span<T>, aka interior pointer, aka stackonly struct at https://github.com/dotnet/roslyn/issues/20127

### CODE REVIEW

**Comment by VSadov** [comment]
@jcouv - FYI

File: build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props [hunk 1]
```diff
@@ -16,5 +16,19 @@
   <PropertyGroup>
     <!-- Compiler server is not currently supported on CoreCLR -->
     <UseSharedCompilation>false</UseSharedCompilation>
+    <CscToolPath>$(MSBuildThisFileDirectory)..\tools\bincore</CscToolPath>
```
Comment (jcouv) [inline build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props]: Are these changes intentional? Why are they showing up in this diff?

Comment (jcouv) [reply build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props]: @VSadov Ping on question above.

Comment (jcouv) [reply build/NuGetAdditionalFiles/Microsoft.NETCore.Compilers.props]: Vlad said that is intentional. #Closed

**Comment by VSadov** [comment]
test windows_debug_unit64_prtest please

**Comment by VSadov** [comment]
looks like windows_debug_unit64_prtest got stuck and did not start at all.

**Comment by VSadov** [comment]
CC: @jaredpar @Pilchie

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler - FYI

**Comment by Pilchie** [comment]
Tagging @dpoeschl too.  Assuming that the manual IDE testing for new language features has been done, I'm okay with merging this.

**Comment by VSadov** [comment]
CC: @MeiChin-Tsai - I think this may need ask mode approval.

**Comment by VSadov** [comment]
@dotnet-bot test windows_release_unit64_prtest please

**Comment by VSadov** [comment]
windows_release_unit64_prtest failed with "cannot access file" during build

**Comment by VSadov** [comment]
test windows_debug_vs-integration_prtest please

**Comment by VSadov** [comment]
test windows_release_vs-integration_prtest please

**Comment by VSadov** [comment]
RE: IDE manual testpass

I went through the IDE scenarios - to be sure.

I used fairly large solution - the entire CoreFxLab repo. 

There was nothing extraordinary. 
I logged some bugs - mostly in cases where refatorings lose "readonly" on "ref readonly". In every case the feature otherwise worked correctly.

There was an assert that seems bogus/ignorable specifically related to "ref readonly" and hard to reproduce. - https://github.com/dotnet/roslyn/issues/22237
It will not have effect on code that does not have "ref readonly" in it or in Release builds, so I think we should look into this after the merge.

Otherwise everything worked fine.

**Comment by VSadov** [comment]
# Testing tags

- :mag: = Scenarios that are regularly tested against internal builds
- :fast_forward: = Scenarios that must work as expected before new language features are merged

When doing a test pass, copy this page and consider using these status indicators:

- :white_check_mark: = Tested & Approved (possibly with linked bugs)
- :x: = Merge blocking
- :construction: = Non-blocking bugs
- :ok: = Issue has been fixed
- :free: = No expected impact
- :question: = Open question

<a name="testfootnote1"><sup>1</sup></a> Feature updated by compiler team, with appropriate unit tests.

<a name="testfootnote2"><sup>2</sup></a> Feature requires more complete testing by IDE team.

# Product Features

### Core Scenarios in Editing, Navigating, and Viewing

| Category | Feature/Description | C# Signoff/Notes | VB Signoff/Notes | F# Signoff/Notes |
| --- | --- | --- | --- | --- | 
| **Enable/Disable** | :fast_forward: **Feature Flags** <br />To completely enables/disable new compiler features in the compiler & IDE | :white_check_mark: | | N/A |
| **Typing** | :fast_forward: **General Typing**<br />- Type and paste new constructs<br />- Nothing interferes with verbatim typing | :white_check_mark: | | |
| | :mag: :fast_forward: **Completion**<br />- Typing new keyword/construct names<br />- Dotting off of new constructs<br />- Matching part of the identifier is highlighted (including word prefix matches) [Visual Studio 2015 Update 1]<br />- Target type preselection [Visual Studio 2017]<br />IntelliSense filtering [Visual Studio 2017] | :white_check_mark: | | |
| | :fast_forward: **Formatting** <br />- Spacing in and around new constructs<br />- Spacing options<br />- Format Document command<br /> `Tools > Options` settings should be respected | :white_check_mark: | | |
| | :fast_forward: **Automatic Brace Completion** (*C# only*) <br />- Auto-insert close brace<br />- Shift+Enter commit of IntelliSense and any pending brace completion sessions (Known issue: https://github.com/dotnet/roslyn/issues/18065) | N/A | | N/A |
| | :fast_forward: **Indentation** <br />- Typing `Enter` in an unfinished statement indents the next line |:white_check_mark: | | |
| **Navigating** | :mag: :fast_forward: **Go To Definition** <br />- F12 from callsites to definition<br />- Ctrl+click [Visual Studio 2017 version 15.4] | :white_check_mark: | | |
| | :fast_forward: **Go To Implementation** <br />- Ctrl+F12 to jump from virtual members to their implementations<br />- Jump from inheritable types to their implementations |:white_check_mark: | | N/A |
| | :mag: :fast_forward: **Find All References** <br />- Lists references to a symbol in "Find Symbol Results" window<br />- Shows results in hierarchy grouped by definition [Visual Studio 2015]<br />- Results should be groupable/filterable/classified appropriately [Visual Studio 2017] <br />- Find All References on literals [Visual Studio 2017 version 15.3] | :white_check_mark:| | |
| **Viewing** | :mag: :fast_forward: **Colorization** <br />- Keywords, literals, and identifiers colored appropriately in code<br />- Colors should theme appropriately<br />- The "Blue Theme (Additional Contrast)" should have sufficiently dark colors |:white_check_mark: | | |
| | :fast_forward: **Error Squiggles** <br />- Squiggles appear as expected on reasonable spans |:white_check_mark: | | |

### Code Transformations
| Category | Feature/Description | C# Signoff/Notes | VB Signoff/Notes | F# Signoff/Notes |
| --- | --- | --- | --- | --- |
| **Refactoring with UI** | :mag: **Inline Rename (with UI)**<br />- Dashboard shows correct information<br />- Highlighted spans are updated appropriately<br />- Rename operation updates the correct set of symbols |:white_check_mark:  | | |
| | :mag: **Change Signature (with UI)**<br />- Updates all direct & cascaded definitions/callsites<br />- Shows appropriate signature & parameter previews in UI<br />- Reorder and Remove in the same session [Visual Studio 2015] | :white_check_mark: https://github.com/dotnet/roslyn/issues/22239| | N/A |
| | :mag: **Extract Interface (with UI)**<br />- Generated Interface has expected shape<br />- UI shows appropriate method previews | :white_check_mark: https://github.com/dotnet/roslyn/issues/22238 | | N/A |
| | **Generate Type (with UI)**<br />- Dialog gives all valid options<br /> | N/A | | N/A |
| | **Generate Overrides** [Visual Studio 2017 version 15.3] | N/A | | N/A |
| **Refactorings** | **Rename Tracking**<br />- Tracking span tracks & dismisses as expected<br />- Invokable from references [Visual Studio 2015] |:white_check_mark: | | N/A |
| | :mag: **Extract Method**<br />- Extracted method has the expected signature<br />- All arguments/return values handled correctly<br />- Extracted code block is reasonable<br />- Automatically starts Inline Rename | N/A | | N/A |
| | **Introduce Variable**<br />- Introduced variable has the expected signature and initializer expression<br />- "Introduce for All" correctly finds dupes |N/A | | N/A |
| | **Inline Temporary Variable**<br />- Inlined values are appropriately expanded/reduced |N/A | | N/A |
| | **Organize Usings**<br />- Honors "Place 'System' namespace first" option |N/A | | N/A |
| | **Convert `Get*` Methods to Properties**<br />*Add Description* |N/A | | N/A |
| | :mag: **Encapsulate Field**<br />- Select a field and convert it to a property backed by the field<br />- Try selecting multiple fields at once |N/A | | N/A |
| | **Modifier Ordering** [Visual Studio 2017 version 15.5] |:white_check_mark: | | N/A |
| | **Convert `ReferenceEquals` to `is null`** [Visual Studio 2017 version 15.5] | N/A | | N/A |
| | **Add Missing Modifiers** [Visual Studio 2017 version 15.5] | :white_check_mark: https://github.com/dotnet/roslyn/issues/22240 | | N/A |
| | **Convert Lambda to Local Function** [Visual Studio 2017 version 15.5] | :white_check_mark: https://github.com/dotnet/roslyn/issues/22241 | | N/A |
| | **Move Declaration Near Reference** [Visual Studio 2017 version 15.5] |N/A | | N/A |
| | **Introduce Pattern Matching** [Visual Studio 2017 version 15.5] | N/A | | N/A |
| | **Simplify Inferred Tuple Names** [Visual Studio 2017 version 15.5] | _Requires C# 7.1 or greater_ |N/A | N/A |
| | **Convert keyword and symbol references to doc comment tags** [Visual Studio 2017 version 15.5] | | | N/A |
| **Fixes** | **Add Using**<br />- Triggers on appropriate constructs<br />- Including NuGet references<br />- Including Referenced Assemblies<br />- Includes spelling fixes | N/A | | |
| | **Generate Local**<br />- Select an expression and introduce a local variable to represent it<br />- This should start an Inline Rename session |N/A | | N/A |
| | **Generate Field**<br />- Select an expression and introduce a field to represent it<br />- This should start an Inline Rename session | N/A | | N/A |
| | **Generate Method/Constructor**<br />- Call a nonexistent method or constructor to generate it from its usage<br />- Generated method has the expected signature and accessibility<br />- Add parameter to existing constructor from callsite [Visual Studio 2017 version 15.3] |N/A | | N/A |
| | **Generate Constructor from members**<br />- Select fields/properties to generate a constructor accepting corresponding arguments<br />- Generated constructor has the expected signature and accessibility |N/A | | N/A |
| | **Implement Interface**<br />- Only missing methods added<br />- All added methods have the expected signature and accessibility | :white_check_mark: https://github.com/dotnet/roslyn/issues/22242 | | N/A |
| | **Implement IDisposable**<br />- Implement IDisposable and you should see a large block of code properly implementing that particular interface |N/A | | N/A |
| | **Implement Abstract Class**<br />- Inherit from an abstract class, and you should be able to auto-generate all of the missing members |:white_check_mark: N/A | | N/A |
| | **Remove Unused Variables** [Visual Studio 2017 version 15.3]| N/A | | N/A |
| | **Remove Unused Usings** |N/A | | |
| | **Sort Usings** |N/A | | N/A |
| | **Convert Get Methods to Properties**<br />- Name a method `GetStuff` and convert it to a property called `Stuff` |N/A | | N/A |
| | **Make Method Async/Sync**<br />- Add an `await` to a synchronous method, you should be offered to add the async keyword<br />- Remove all `await` keywords from an async method, you should be offered to remove the async keyword |N/A | | N/A |
| | **Use Object Initializer Over Property Assignment**<br />- Create a new instance of a type and then assign each of its properties on subsequent lines<br />- You should be offered to convert that to an object initializer |N/A | | N/A |
| | **Insert Digit Separators** [Visual Studio 2017 version 15.3] |N/A | | N/A |
| **Code Gen** | :mag: **Snippets**<br />- Tab completion, presence in completion list<br />- Insertion via Snippet Picker UI (Ctrl + K, Ctrl + X) or (Ctrl + K, Ctrl + S)<br />- (VB) Snippet Picker UI via `?<Tab>`<br />- (VB) Special snippet completion list (`p?<esc><tab>`) |N/A | | N/A |
| | **Event Hookup on Tab** (*C# only*)<br />- Type "+=" after an event name and QuickInfo shows<br />- Invoking should pick good name & launch Inline Rename |N/A | N/A | N/A |
| | **End Construct Generation** (*VB only*)<br />- Type `Sub Test()` and hit enter, the `End Sub` should be generated automatically | N/A | | N/A |
| | **Automatic End Construct Update** (*VB only*)<br />- Type `Sub Test()` and `End Sub`, changing `Sub` to `Function` in either one should update the other | N/A | | N/A |
| | **Spell checking**<br />- Type a name that's close to a variable name but off by a character or two<br />- Lightbulb should have option to fix up the spelling<br />- Includes type names that will require a using |N/A | | N/A |
| | **Move type to file**<br />- Lightbulb to move type to another file when the type name doesn't match the filename<br />- Option to change the file name if the type doesn't match the file name |:white_check_mark: | | N/A |
| | **Convert between properties and Get methods** <br />- Offers to change a method named `GetStuff` to a property named `Stuff` |N/A | | N/A |
| | **Convert auto property to full property** [Visual Studio 2017 version 15.5] | N/A| | N/A |
| | **Add missing cases**<br />Use a `switch` on a strict subset of an Enum's members<br />- It should offer to generate the rest of the cases |N/A | | N/A |
| | **Add null checks for parameters** [Visual Studio 2017 version 15.3] |N/A | | N/A |
| | **Change base for numeric literals** [Visual Studio 2017 version 15.3] |N/A | | N/A |
| | **Convert if to switch** [Visual Studio 2017 version 15.3] |N/A | | N/A |
| | **Resolve git merge conflicts** [Visual Studio 2017 version 15.3] |N/A | | N/A |
| | **Add argument name** [Visual Studio 2017 version 15.3 |:white_check_mark: | | N/A |
| | **Fade and remove unreachable code** [Visual Studio 2017 version 15.5] |N/A | | N/A |
| | **Add missing file banner** [Visual Studio 2017 version 15.5] |N/A | | N/A |

### IDE Features
| Category | Feature/Description | C# Signoff/Notes | VB Signoff/Notes | F# Signoff/Notes |
| --- | --- | --- | --- | --- |
| **General** | **Signature Help**<br />- Overloads shown with appropriate, colorized signature |:white_check_mark: | | |
| | **Quick Info**<br />- Hover on identifiers<br />- On completion list items |:white_check_mark: | | |
| | :mag: **Outlining**<br />- Make sure outlining is enabled under options<br />- Define a method and expect to see a collapsible region around the method definition<br />- Make sure collapse and expand work |:white_check_mark: | | |
| | **Brace Matching** (*C# only*)<br />- Highlights matching brace token, if caret is on an open or close brace.<br />- Hovering on a close brace shows the code around the open brace | :white_check_mark:| N/A | N/A |
| | **Highlight References**<br />- Ensure "Highlight references to symbol under cursor" is enabled in options<br />- If caret is on an identifier, all references to that identifier in the active document should be highlighted<br />- We also show related keywords, so placing the caret on a return should show the other returns, if should show elses, try shows catches and finallys, etc.<br />- Should be able to navigate between highlighted references with Ctrl+Shift+Up/Down |:white_check_mark: | | |
| | :mag: **Peek**<br />Press Alt + F12 after placing the cursor on a predefined Type or predefined member and expect to see to a temporary window showing the appropriate definition |:white_check_mark: | | |
| | :mag: **Navigation Bars**<br />- Open some existing source files, make sure you can navigate around the file choosing classes or methods.<br />- Switch between project contexts<br />- In VB, the NavBar can do code generation for events and for New/Finalize methods |:white_check_mark: | | |
| | **Metadata As Source**<br />- Press F12 on a predefined type and expect the cursor to move the predefined type definition inside a Metadata-As-Source Generated document.<br />- Expect to see the xml doc comments collapsed above the method. |:white_check_mark: | | N/A |
| | :mag: **Navigate To**<br />- Place caret on a user defined Type reference and press "ctrl + ,"<br />- This should list the User Defined Type in the drop down on the Upper Right corner of the editor and selecting this item will move the cursor to the User Reference Definition<br />- Filters per kind of symbol [Visual Studio 2017] |:white_check_mark: | | |
| | **Go to Next/Previous Method**<br />- `Edit.NextMethod` and `Edit.PreviousMethod` should work<br />- You may need to set up keyboard bindings for these commands under `Tools > Options > Environment > Keyboard` |:white_check_mark: | | N/A |
| | **Solution Explorer Pivots**<br />- Define a Type and some members in a sample Document.<br />- Expand the Document listed in the Solution Explorer window and expect to see the Type and Members defined<br />- Right-click types and try Base Types / Derived Types / Is Used By / Implements<br /> - Right-click methods and try Calls / Is Called By / Is Used By / Implements |:white_check_mark: | | N/A |
| | **Call Hierarchy**<br />- Place the caret over a method, right click & select View Hierarchy<br />- This should open the "Call Hierarchy window" listing the methods that call the original method and also the callsites within each calling method. | :white_check_mark:| | N/A |
| | :mag: **Code Lens**<br />- Make sure Code Lens is enabled from the options. Look for an adornment on top of each method declaration with lists the number of references for that method, last time someone modified the method, who modified the method and other information | :white_check_mark:| | N/A |
| | **Project System**<br />- Open/close a project<br />- Add/remove references<br />- Unload/reload projects<br />- Source Control integration (adding references checks out projects, etc.) | N/A| | |
| | **Debugger IntelliSense**<br />- Hit a breakpoint (or step) and verify that there's IntelliSense in the Immediate Window (C#, VB)<br />- Type an expression, and hit enter. Verify it's evaluated. Type another expression. IntelliSense should still work.<br />-    (VB) there should be IntelliSense if you type "?" followed by an expression (eg, the text of the line in the window is "?foo")<br />- Hit a breakpoint (or step) and verify that there's IntelliSense in the Watch Window<br />- Hit a breakpoint (or step) and verify that there's IntelliSense in the Quick Watch Window<br /> - Verify intellisense in the Conditional Breakpoint view<br />- Verify each of the above scenarios after hitting f5 and hitting another breakpoint, and after stepping |:white_check_mark: | | |
| | **Breakpoint Spans**<br />- The span highlighted when a breakpoint is set should be logical |:white_check_mark: | | |
| | **Code Model / Class Designer**<br />- Right click a file in Solution Explorer & choose View Class Diagram.<br />- This shows the "Class Details" window where you can add/remove members, change their type/accessibility/name, parameter info, etc. | | | N/A |
| | **Object Browser / Class View**<br />- Open object browser and classview and verify that project contents are represented live.<br />- Should be able to invoke find all references, navigate to the definition, and do searches. | :white_check_mark:| | N/A |
| | **Lightbulb**<br />- Should work with `Ctrl+.` and the right-click menu<br />- Should include a diff view<br />- Should include options to fix all in Document/Project/Solution, and to Preview Changes<br />Varying icons for lightbulb items [Visual Studio 2017 version 15.4] |:white_check_mark: | | |
| | **Line Separators**<br />- Turn the option on under `Tools > Options`<br />- Ensure there's a line between methods. |:white_check_mark: | | N/A |
| | **Indent Guides**<br />- Vertical dotted lines appear between braces for declaration-level things<br />- Hovering on the line shows context |:white_check_mark: | | |
| **Code Style** | **Naming Rules** |N/A | | N/A |
| | **Use this./me.** |N/A | | N/A |
| | **Use predefined type** |N/A | | N/A |
| | **Prefer object/collection initializer** |N/A | | N/A |
| | **Prefer explicit tuple name** |N/A | | N/A |
| | **Prefer coalesce expression over null check** |N/A | | N/A |
| | **Prefer null propagation over null check** |N/A | | N/A |
| | **var preferences** (*C# only*) |N/A | N/A | N/A |
| | **Prefer braces** (*C# only*) |N/A | N/A | N/A |
| | **Prefer pattern matching over is/as checks** (*C# only*) |N/A | N/A | N/A |
| | **Use expression body** (*C# only*) |N/A | N/A | N/A |
| | **Prefer inlined variable declaration** (*C# only*) |N/A | N/A | N/A |
| | **Prefer throw expression** (*C# only*) |N/A | N/A | N/A |
| | **Prefer conditional delegate call** (*C# only*) |N/A | N/A | N/A |

**Comment by VSadov** [comment]
According to Jenkins the integration tests have been failing continuously for the last 7 days now. They are not merge-blocking though.

I will ignore the integration test results.

**Comment by MeiChin-Tsai** [comment]
Please make sure all tests pass. Otherwise, approved.

**Comment by VSadov** [comment]
Thanks!! All required tests are passing.
